### ISSUE TITLE
mullvad app custom DNS not accepting 127.0.0.1

### ISSUE THREAD

**Issue by ghrnhge644** [mullvad/mullvadvpn-app#4033]
## Issue report

Windows 10

App version: 2022.3 2022.4 2022.5beta-1

### Issue description

In version 2022.2 and previously when you set 127.0.0.1 in the custom DNS section of the mullvad app it would also be in the Mullvad network adapter. But after this version and onward the ip 127.0.0.1 is not being set on the mullvad network adapter even if it's in the mullvad app dns section. But other ips will work (tried with mullvad's local dns 100.64.0.3). This is an issue because I am trying to route my DNS to 127.0.0.1 to use portmaster. I can manually go on the mullvad network adapter and set 127.0.0.1 and it will work but this adapter will reset each time I change server or restart my computer. Doing this every day becomes really tedious. Some users already reported the issue #3842 . I also have added 127.0.0.1 on all my other network adapter but it doesn't work (DNS requests take 10+seconds). Only thing that worked was adding 127.0.0.1 to the mullvad network adapter but like I said it becomes tedious to do this every single time.

**Comment by dlon** [comment]
We're considering fixing this by relaxing some restrictions on loopback traffic.

In the meantime, you should be able to work around the issue by setting the custom DNS servers to `127.0.0.1` in the app, and setting the *primary* DNS server to `127.0.0.1` on your real network adapter. Leave the `Mullvad` DNS server list empty (as it is by default).

If the above still does not work, maybe the DNS client service is trying to use IPv6. You might want to follow all of the same steps for the IPv6 interfaces, using `::1` instead of (or in addition to) `127.0.0.1`.

If the above still fails, you could override all of the interface-specific DNS servers by adding an NRPT rule instead. To do so, run the following in PowerShell as admin:

```
Add-DnsClientNrptRule -Namespace "." -NameServers "127.0.0.1"
```

Hope this helps.

**Comment by ghrnhge644** [comment]
> We're considering fixing this by relaxing some restrictions on loopback traffic.
> 
> In the meantime, you should be able to work around the issue by setting the custom DNS servers to `127.0.0.1` in the app, and setting the _primary_ DNS server to `127.0.0.1` on your real network adapter. Leave the `Mullvad` DNS server list empty (as it is by default).
> 
> If the above still does not work, maybe the DNS client service is trying to use IPv6. You might want to follow all of the same steps for the IPv6 interfaces, using `::1` instead of (or in addition to) `127.0.0.1`.
> 
> If the above still fails, you could override all of the interface-specific DNS servers by adding an NRPT rule instead. To do so, run the following in PowerShell as admin:
> 
> ```
> Add-DnsClientNrptRule -Namespace "." -NameServers "127.0.0.1"
> ```
> 
> Hope this helps.

I already tried your 2 first steps but they did not work for me.

The last step worked for me (NRPT rule) thank you! hopefully it persists after reboot. I will be reporting back any issues I have for this issue.

**Comment by dlon** [comment]
Fixed by #4047.
